### PR TITLE
[Feature] GroupSwitchModal miscellaneous cleanup

### LINKED ISSUE #1210: [Feature] Group switching
### 🧩 Summary

UI designs for participants to switch to a new group. (mobile and desktop).

You do not need to code up any of the workflows for executing the switch. All the workflows are completed within Airtable. This interface allows participants to effectively create a record in the Group Switching table from the Course Runner base.

### ✅ Definition of done

- When users click "Can't make it?" from the discussion info box on the course unit page, it opens this modal
- When users click on the unit dropdown, it shows the list of all units for the course
- When users click the action dropdown, it shows 2 actions. (I'm deprioritising giving the option to "Join group for one unit" in this sprint)
- When users click the action dropdown, the action which has already been selected has a checkmark next to it
- When users have selected an action and a unit, the list of groups available should filter the groups based on the groups which match the "Who can switch into this group?" field and the participant's "Bucket". It should be arranged chronologically. The number of spots available is the difference between the "Max participants per group" field from the Round table and the number of participants in the group.
- If a group has equal to or more than the max participants per group, participants should not be able to select it
- When users select a group and the click "Confirm", the data is sent to the Course Runner table for Group switching with the "Request status" = Requested. The data for the participants past and new discussion and group should be logged.
- When users submit a manual request, the data is sent to the Course Runner table for Group switching with the "Manual request?" checkbox field checked and the "Request status" field = Resolve. The participant's current discussion should be logged.
- Only participants for discussions should be able to access this, _not facilitators_

### 📌 Notes / Designs

https://www.figma.com/design/6PJJZThivMuDTZ9xcoWvBs/Group-switching-UI?node-id=3037-441&t=s6zDvjAAKGInxNQz-0

### Suggested Priority Level

{"P1" => "High"}

### PR DESCRIPTION
**Pull request by Will-Howard**
# Description

The last few bits of cleanup to match the designs in the original group switching issue (#1210):
- Add info about the group being switched into to the "Success!" screen
- Add "Update availability" to the manual switching screen
- When switching permanently, use the recurring time on the group rather than the time of the next discussion. Also display this as e.g. "Wed 11:00 AM", rather than "Nov 13 11:00 AM"
- Add icons to the title of "Success!" and "We are working on your request" screens
- Don't allow switching into discussions that have already passed
- I also added some tests for the calculateGroupAvailability function used on the backend

## Issue

Fixes #1210

## Developer checklist

- [x] Front-end code follows [Component Accessibility Checklist (for Testing)](https://github.com/bluedotimpact/bluedot/blob/master/DEVELOPMENT_HANDBOOK.md#component-accessibility-checklist)
- [x] [Lacking, see #1372] Considered having snapshot tests and/or happy path functional tests#13
- [x] Considered adding Storybook stories

## Screenshot

| 📸 |  |
|---------|---|
| 📱  | Screenshot 2025-09-23 at 09 03 36 |
| | Screenshot 2025-09-23 at 09 03 05 |
| | Screenshot 2025-09-23 at 09 03 19 |
| | Screenshot 2025-09-23 at 09 04 36 |
| 🖥️ | Screenshot 2025-09-23 at 08 58 36 |
| | Screenshot 2025-09-23 at 08 59 01 |
| | Screenshot 2025-09-23 at 09 01 13 |
| | Screenshot 2025-09-23 at 09 00 58 |

### CODE REVIEW

File: apps/website/src/components/courses/GroupSwitchModal.tsx [hunk 1]
```diff
@@ -71,24 +73,29 @@ const getGroupSwitchDescription = ({
...
   }
-
-  if (isSelected) {
-    return <span className="text-[#0037FF]">You are switching into this group for all upcoming units</span>;
+  if (spotsLeft === null) {
+    return <><UserIcon className="-translate-y-px" /><span>Spots available</span></>;
   }
-
-  return undefined;
+  return <><UserIcon className="-translate-y-px" /><span>{spotsLeft} spot{spotsLeft === 1 ? '' : 's'} left</span></>;
```
Comment (marn-in-prod) [inline apps/website/src/components/courses/GroupSwitchModal.tsx:98]: Nit: Can we refactor this for clarity?

It's not semantically clear why spotsLeft being null and hasAnySpotsLeft being true means that there are spots available. Maybe just rename spotsLeft to `spotsLeftIfKnown` or something similar?

File: apps/website/src/components/courses/GroupSwitchModal.tsx
Comment (marn-in-prod) [inline apps/website/src/components/courses/GroupSwitchModal.tsx:1]: Nit: When a user successfully changes which group their in, can they make another group switch immediately or do they need to refresh the page (since showSuccess becomes True and I didn't notice a way for it to be false)?

Not something you really need to fix, unless you think you can do it easily since I think it's probably a rare-ish edge case.

**Review by marn-in-prod — APPROVED** [review]
Generally looks good. Although having a few tests to confirm that users can switch groups would be great since this component is complicated

File: apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts [hunk 2]
```diff
@@ -89,45 +85,40 @@ function calculateGroupAvailability({
       discussionsByUnit[unitKey] = [];
     }
 
+    const hasStarted = discussion.startDateTime * 1000 <= now;
+
     discussionsByUnit[unitKey].push({
       discussion,
       spotsLeft,
```
Comment (joshestein) [inline apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts:92]: For groups you've made `spotsLeft` depend on whether `hasStarted` is true or not. Do we not need to make that distinction for discussions?

File: apps/website/src/components/courses/GroupSwitchModal.tsx [hunk 3]
```diff
@@ -410,22 +447,36 @@ const GroupSwitchModal: React.FC<GroupSwitchModalProps> = ({
...
+                    url={`https://availability.bluedot.org/form/bluedot-course?email=${encodeURIComponent(auth.email)}&utm_source=bluedot-group-switch-modal`}
+                    aria-label="Update availability (open in new tab)"
+                  >
+                    Update availability
+                  </CTALinkOrButton>
+                </div>
+              )}
+              <div className="flex gap-2 justify-end">
+                <CTALinkOrButton
+                  className="mx-auto"
```
[the remark this answers is not in the document]
Comment (joshestein) [reply apps/website/src/components/courses/GroupSwitchModal.tsx:471]: @herrhaase maybe you can comment? I'm specifically referring to the first button being bottom-left aligned and the second set of buttons being centered.

image

and

image

File: apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts [hunk 4]
```diff
@@ -89,45 +85,40 @@ function calculateGroupAvailability({
...
-          existing.nextDiscussionStartDateTime = discussion.startDateTime;
-        } else {
-          existing.nextDiscussionStartDateTime = Math.min(
-            existing.nextDiscussionStartDateTime,
-            discussion.startDateTime,
-          );
+      if (!hasStarted) {
+        // Update spotsLeft
+        if (existing.spotsLeft !== null && spotsLeft !== null) {
+          // Clamp spotsLeft to be >= 0
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply apps/website/src/pages/api/courses/[courseSlug]/group-switching/available.ts:118]: You're right, it was already clamped above so this was redundant

File: apps/website/src/components/courses/GroupSwitchModal.tsx [hunk 5]
```diff
@@ -71,24 +73,29 @@ const getGroupSwitchDescription = ({
...
+  const hasAnySpotsLeft = spotsLeft !== 0;
+  if (!hasAnySpotsLeft) {
+    return <><UserIcon className="-translate-y-px" /><span>No spots left</span></>;
   }
-
-  if (isSelected) {
-    return <span className="text-[#0037FF]">You are switching into this group for all upcoming units</span>;
+  if (spotsLeft === null) {
+    return <><UserIcon className="-translate-y-px" /><span>Spots available</span></>;
   }
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply apps/website/src/components/courses/GroupSwitchModal.tsx:97]: Yep it does, this is for the case where `maxParticipants` isn't set on the round, so we assume you're always allowed to switch into the group (this case does actually appear in the airtable). Martin also flagged this above so I'll use his suggestion of renaming to `spotsLeftIfKnown`
